ratbag_emu/protocol/hidpp20.py

```diff
@@ -190,3 +192,50 @@ def IFeatureSet(self, data, ase, args):
 
     def IFeatureInfo(self, data, ase, args):
         return
+
+    def DeviceInformation(self, data, ase, args):
+        # entityCnt, unitId, transport, modelId, extendedModelId = getDeviceInfo()
+        if ase == 0:
+            reply = [
+                # entityCnt
+                len(self.entities),
+                # unitId
+                random.randint(0x00, 0xFF),
+                random.randint(0x00, 0xFF),
+                # transport
+                0x00,
+                0x06,  # USB/eQUAD
+                # modelId (FIXME: Add modelId value)
+                0x00,
+                0x00,
+                # extendedModelId (should not matter, we will keep it at 0)
+                0x00,
+            ]
+            logger.debug(f'FIXME: add modelId')
+            logger.debug(f'getDeviceInfo() = (entityCnt) {reply[0]}, (unitId) {pack_be_u16(reply[1], reply[2])}, '
+                '(transport) USB/eQUAD, (modelId) {pack_be_u16(reply[5], reply[6])}, (extendedModelId) {reply[7]}')
```

Comment:
@whot can you comment on this please?

Reply:
github isn't showing me anything here, it looks like this is still the same commit?

This PR is getting out of hand tbh, i've got no overview anymore what's happening where here...

Reply:
```python
        if ase == 0:
            reply = [
                # entityCnt
                len(self.entities),
                # unitId
                0xAA,
                0xBB,
                # transport
                0x00,
                0x06,  # USB/eQUAD
                # modelId (FIXME: Add modelId value)
                0x00,
                0x00,
                # extendedModelId (should not matter, we will keep it at 0)
                0x00,
            ]
            logger.debug(f'FIXME: add modelId')
            self._log_msg('getDeviceInfo',
                          out={
                              'entityCnt': reply[0],
                              'unitId': pack_be_u16(reply[1], reply[2]),
                              'transport': 'USB/eQUAD',
                              'modelId': pack_be_u16(reply[5], reply[6]),
                              'extendedModelId': reply[7]
                          })
            self.protocol_reply(data, reply)
```

I think writing structures for all packet types is a bit overkill, even though it would look better here.

Reply:
`_log_msg()` is defined as:

```python
    def _log_msg(self, function, input={}, out={}):
        def format_args(args):
            return ', '.join(f'({key}) {value}' for key, value in args.items())

        if 'error' in out:
            logger.debug(f'{function}({format_args(input)}) = {out["error"]}')
        else:
            logger.debug(f'{function}({format_args(input)}) = {format_args(out)}') 
```

Reply:
@whot can you just give a word on this? Thanks :D

Reply:
> I think writing structures for all packet types is a bit overkill, 

But that's what you end up doing anyway with the above. As soon as you need to pack/unpack and access the various `reply` indices you're effectively writing struct parsers.

But yeah, the above is a lot better than the original version. 

I'm a bit confused by the `error` though. Can we pass this in as `error=None` argument instead of hiding it in the `out` dictionary?

Reply:
> But that's what you end up doing anyway with the above. As soon as you need to pack/unpack and access the various reply indices you're effectively writing struct parsers.

Yes but the code will not be reused so I fail to the value in having separated structure definitions. Either way I have no problem changing that as long as the structure definition stays close to the feature implementation, but I am going to leave that as it is for now to try to finally get this merged.

> I'm a bit confused by the error though. Can we pass this in as error=None argument instead of hiding it in the out dictionary?

That makes more sense.